### DISCUSSION TITLE
URQL with Firebase

### DISCUSSION THREAD

**Discussion by Cazineer** [urql-graphql/urql#2908] [Q&A] [answered]
I'm attempting to use urql with SvelteKit and Firebase.

I'm using a Svelte Store to store the user, which gets updated whenever the auth state changes via Firebase's `onAuthStateChanged` listener.

I've also put the URQL client in a Svelte Store so that whenever the user updates due to an auth state change, a new client can be created with the updated auth credentials. I know the docs recommend using  context.

I did not see a way with the URQL client to trigger an `authState` update any other way (incrementally add the auth headers)? It appears from the documentation that in order to update authState (outside generating a new client), URQL must call `getAuth` and update the auth. _This does not appear to support async._

I guess I'm struggling to get a good architecture here and would prefer to not have to generate a new URQL client every time there is an auth state change? With `graphql-request `for example, I can simply call `setHeaders` on the main client whenever there is an auth state change.

Thanks!

**Comment by JoviDeCroock** [comment, marked as answer]
Hey and welcome!

I have answered this before [here](https://github.com/urql-graphql/urql/discussions/2647#discussioncomment-3501464) basically what I do is I register an event-emitter in the `authExchange` when my `authState` changes I dispatch an event and mutate the `authState` variable.

[This discussion has a completer code example](https://github.com/urql-graphql/urql/discussions/2855)

Recreating a client is a good option as well but it does clear the cache every time which could be unwanted, I do advise on maybe re-creating the client on logout if authenticated data remaining in the in-memory cache is an issue.

**Reply by Cazineer** [reply]
Thanks. I can see this working but it does feel quite weird that there isn't a manual setAuth or updateAuth on the client. 

From the example you linked is `AuthState` a URQL type? I've scoured the docs/Google and did not find it and as thus, I'm stalled with even getting the referenced code to work. 

Side note, I'm super proficient with Rust, Go, C and C++ and I detest TypeScript. I've never struggled with a type system in any language but the sheer amount of time I loose with TypeScript, trying to understand its garbage error messages, or trying to hunt down some type, from some lib feels just inane. 

Even trying to work with the simple examples in the URQL docs, TS just melts down and provides absolute gibberish to help the user.

**Reply by JoviDeCroock** [reply]
Well the client is agnostic to the exchanges so it's quite generic. The type was one created by the user not from urql.

I can imagine the loose nature of TS and the cryptic warnings being an issue

**Reply by Cazineer** [reply]
Thanks for your quick help/responses.

This is where I ended up after looking at your links. Is there anything you'd change, etc?

https://gist.github.com/danielcasler/29a5200f0521feb215364b4f144912f5

**EDIT**:
 
In case others find this, here is a Gist of my Firebase user store, which sets the `onAuthStateChanged` observer end emits the token whenever the auth state changes in Firebase.

https://gist.github.com/danielcasler/7ec4d746591dc0daa2dc4fd611cc3fb9

**Reply by Cazineer** [reply]
Although this partially works, I'm now running into issues where if a page is reloaded and that page has a query that requires authentication, the query will run before Firebase auth has initialized, which then runs the `onAuthStateChanged` observer. Because the observer hasn't run yet, the token hasn't emitted, which means `authState` has been updated. I've tried adding a `known` promise to the user store here:

https://gist.github.com/danielcasler/7ec4d746591dc0daa2dc4fd611cc3fb9

I've then attempted to await the `known` promise and unpause (query was set to pause to start) a `queryStore` upon completion but I don't know if a paused query can have it's `authState`updated or if the query is still running before the `authState` gets updated, but it doesn't work as I would expect. 

Maybe I'm just tired but it doesn't seem like urql is very friendly to BaaS providers such as Firebase, Supabase, Amplify, etc.

**Reply by JoviDeCroock** [reply]
Not really a svelte expert but coming from a React background, setting stuff is synchronous which means that as soon as you call getToken and unpause the query the authState is good to go.

It's generally very hard to facilitate for platforms like firebase/... but most of the solutions I have seen worked. It's normal that you have to provide some glue in these scenario's. I'm terribly sorry but we don't have the time to support everything out there, maybe we could do a better job at documenting all of these cases but still...

Debugger/console can tell you what's happening here in terms of urql land.

**Reply by Cazineer** [reply]
I didn't mean any offence. I certainly appreciate the effort you guys have put into urql. My statement was more coming from that standpoint that auth services like Firebase, Supabase, Auth0, Amplify and many others are fairly common these days and having a way to use them with urql seems like it would be important. 

The issue is definitely using the event emitter. If the client had a method to update authState this would be a non issue because once Firebase was initialized, the authState could be updated on the client directly without relying on an asynchronous event emitter. 

Do you know if a client with a paused query can in fact have its authState updated?

**Reply by JoviDeCroock** [reply]
Example of cognito: https://gist.github.com/mrflip/1e9755a2d99dee197b2bd95e4ec67fe6

As said before we purposefully avoid the client being aware of these exchanges to be generic, this bolsters an ecosystem/...

The authState is purely part of the authExchange this means that the state of the query paused/... has no effect on the state of auth. The client is a global thing that will just dispatch events as they come in, the events run through the exchange pipeline and return a result. So a query/client/... has no authState only that exchange
